arrows/ffmpeg/ffmpeg_video_input.cxx

```diff
@@ -0,0 +1,834 @@
+      //// grab the metadata from this packet if from the metadata stream
+      //else if (this->packet.stream_index == this->data_index)
+      //{
+      //  is_->metadata_.insert(is_->metadata_.end(), is_->packet_.data,
+      //    is_->packet_.data + is_->packet_.size);
+      //}
+
+      if (!got_picture)
+      {
+        av_free_packet(&this->f_packet);
+      }
+    }
+
+    // From ffmpeg apiexample.c: some codecs, such as MPEG, transmit the
+    // I and P frame with a latency of one frame. You must do the
+    // following to have a chance to get the last frame of the video.
+    if (!got_picture)
+    {
+      av_init_packet(&this->f_packet);
+      this->f_packet.data = nullptr;
+      this->f_packet.size = 0;
+
+      int err = avcodec_decode_video2(this->f_video_encoding,
+        this->f_frame, &got_picture,
+        &this->f_packet);
+      if (err >= 0)
+      {
+        this->f_pts += static_cast<int64_t>(this->stream_time_base_to_frame());
+      }
+    }
+
+    // The cached frame is out of date, whether we managed to get a new
+    // frame or not.
+    this->current_image_memory = nullptr;
+
+    if (!got_picture)
+    {
+      this->f_frame->data[0] = NULL;
+    }
+
+    return got_picture != 0;
+  }
+
+  // ==================================================================
+  /*
+  * @brief Get the current timestamp
+  *
+  * @return \b Current timestamp.
+  */
+  double current_pts() const
+  {
+    return this->f_pts * av_q2d(this->f_video_stream->time_base);
+  }
+
+  // ==================================================================
+  /*
+  * @brief Returns the double value to convert from a stream time base to
+  *  a frame number
+  */
+  double stream_time_base_to_frame() const
+  {
+    if (this->f_video_stream->avg_frame_rate.num == 0.0)
+    {
+      return av_q2d(av_inv_q(av_mul_q(this->f_video_stream->time_base,
+        this->f_video_stream->r_frame_rate)));
+    }
+    return av_q2d(
+      av_inv_q(
+        av_mul_q(this->f_video_stream->time_base, this->f_video_stream->avg_frame_rate)));
+  }
+
+  bool is_valid() const
+  {
+    return this->f_frame && this->f_frame->data[0];
+  }
+
+  // ==================================================================
+  /*
+  * @brief Return the current frame number
+  *
+  * @return \b Current frame number.
+  */
+  unsigned int frame_number() const
+  {
+    // Quick return if the stream isn't open.
+    if (!this->is_valid())
+    {
+      return static_cast<unsigned int>(-1);
+    }
+
+    return static_cast<unsigned int>(
+      (this->f_pts - this->f_start_time) / this->stream_time_base_to_frame()
+      - static_cast<int>(this->f_frame_number_offset));
+  }
+
+  // ==================================================================
+  /*
+  * @brief Convert a ffmpeg pixel type to kwiver::vital::image_pixel_traits
+  *
+  * @return \b A kwiver::vital::image_pixel_traits
+  */
+  pixel_format pixel_format_from_ffmpeg(AVPixelFormat ffmpeg_format) const
```

Comment:
I suggest simplifying this quite a bit and removing this function and the `pixel_format` enum.  The VXL vidl library was designed to be able to store these various video frame encodings in raw form.  Most of these are packed formats with different resolutions in the different color channels that `vital_image` cannot support.  In our case we should always ask FFMPEG to convert the image into RGB24 for us.  The only exceptions would be
- `RGBA` - create a 4 channel image 
- `GRAY8` - create a 1 channel image
- `MONOWHITE` and `MONOBLACK` - create a 1 channel boolean image

Note that it is quite rare to find videos encoded with these.  Most videos use some variant of YUV.